resilience4j-spring-boot-common/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/AbstractBulkheadConfigurationOnMissingBean.java

```diff
@@ -104,11 +106,19 @@ public ReactorBulkheadAspectExt reactorBulkHeadAspectExt() {
     public ThreadPoolBulkheadRegistry threadPoolBulkheadRegistry(
         ThreadPoolBulkheadConfigurationProperties threadPoolBulkheadConfigurationProperties,
         EventConsumerRegistry<BulkheadEvent> bulkheadEventConsumerRegistry,
-        RegistryEventConsumer<ThreadPoolBulkhead> threadPoolBulkheadRegistryEventConsumer) {
+        RegistryEventConsumer<ThreadPoolBulkhead> threadPoolBulkheadRegistryEventConsumer,
+        Optional<Customizer<Builder>> compositeThreadPoolBulkheadBuilderCustomizer) {
 
         return threadPoolBulkheadConfiguration.threadPoolBulkheadRegistry(
             threadPoolBulkheadConfigurationProperties, bulkheadEventConsumerRegistry,
-            threadPoolBulkheadRegistryEventConsumer);
+            threadPoolBulkheadRegistryEventConsumer,compositeThreadPoolBulkheadBuilderCustomizer);
+    }
+
+    @Bean
+    @Primary
+    public Customizer<Builder> compositeThreadPoolBulkheadBuilderCustomizer(
```

Comment:
what if the user want to override that via custom bean injection , here were forcing this bean as primary bean all the time , can not we just use @ConditionalOnMissingBean ?

Reply:
This is similar to design of `threadPoolBulkheadRegistryEventConsumer` https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring-boot-common/src/main/java/io/github/resilience4j/bulkhead/autoconfigure/AbstractBulkheadConfigurationOnMissingBean.java#L114-L120

We are collecting all beans of type 'Customizer` present in the Context in `CompositeCustomizer`, composing into a single wrapper bean(delegate pattern https://github.com/resilience4j/resilience4j/pull/761/files#diff-fd0197063f47bb044a0902723f6f248eR13) and then using a single bean in framework internals. If we use `@ConditionalOnMissingBean ` then 
1. everywhere we have to loop through customizers present in context(https://github.com/resilience4j/resilience4j/pull/761/files#diff-bc5195d76b00f5bb9fc1eb3d24534e11R97) 
2. Use Optional in DI because bean could be absent in the context. 

Let me know if you still think we should change this.

Reply:
@ConditionalOnMissingBean means it will be injected all the time as long as no one is overriding it as we do for for all resilience4j managed beans so for your points mentioned above :

- if no overridden bean is introduced , your confitional bean will be injected so u will not face the issues u mentioned above 
- if the user override it , then he have to respect it is own purpose as it is being shown in ur code and u will not face the issues u mentioned as well , same goes for most of other resilience managed beans 

Make sense ? or i am missing something ? please let know and thx a ot for this nice PR 👍

Reply:
Thank you for the explanation!

Application can provide multiple customizer beans so we have to make this code a List<Customizer> instead of single Customizer (we smartly fit delegate pattern using CompositeCustomizer) https://github.com/resilience4j/resilience4j/pull/761/files#diff-118f95f880f6d107c67d522fd9712617R112

Now, this list either contains
1. A composite customizer bean (if no override bean is available in the context)
2. Or list of application provided Customizer bean.

The thing that I don't like here is that if no bean override happen we have a List<Customizer> which has only one bean of type CompositeCustomizer. So, essentially a list within list.

I can make that change but want to call out this scenario. Pls confirm!

Reply:
Yes you are correct  ,  that is why maybe your composite interface should not be of a type of Customizer , your composite delegate should be the one that we use in our internal logic once it has  the list of customizers injected into it either by our bean or the overridden composite if any by the users .

So users :

1. can implement as many customizes as they wish
2. your unique composite delegate bean will collect them via spring List injection 
3. then we pass the composite delegate bean to our internal initialization logic which will enable the list of Customizers access 

Now what i do not feel optimal is we have to stream all customizers per configured instance to execute everyone of them to see if it will match the name parameter .

What if we can extend the Customizer interface by having extra method ex:  instanceName() , then through DI you can have a map where the key will be the instance name and the value will the Customizer instance and should get by name from the map if any and in that case u will not need even if the composite and it will be more optimal as you get only the configured Customizer by name and u will not need to stream all of them every-time per instance .

What do u think ?

Reply:
@romeh I agree to your points